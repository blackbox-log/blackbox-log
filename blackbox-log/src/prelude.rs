//! Minimal set of imports for convenience.
//!
//! [`File`](`crate::File`) is not included due to avoid collisions with other
//! types.

pub use crate::frame::{Frame as _, FrameDef as _};
pub use crate::{DataParser, FieldFilter, Headers, ParserEvent};
