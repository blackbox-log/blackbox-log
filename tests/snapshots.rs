use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
