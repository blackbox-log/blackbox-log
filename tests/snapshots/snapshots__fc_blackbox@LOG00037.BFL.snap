---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/fc-blackbox/src/test-data/LOG00037.BFL"
---
FileSnapshot(
  count: 1,
  logs: [
    Ok(LogSnapshot(
      headers: Headers(
        firmware_revision: "Betaflight 4.2.0 (8f2d21460) STM32F745",
        firmware: Betaflight("4.2.0"),
        board_info: Some("HBRO KAKUTEF7"),
        craft_name: Some("AR8"),
        debug_mode: GyroScaled,
        disabled_fields: [],
        features: [
          RxSerial,
          MotorStop,
          Gps,
          Osd,
          AirMode,
          AntiGravity,
        ],
        vbat_reference: Some(2277),
        acceleration_1g: Some(2048),
        gyro_scale: Some(0.017453292),
        min_throttle: Some(1070),
        motor_output_range: Some(MotorOutputRange(
          min: 158,
          max: 2047,
        )),
        unknown: {
          "dyn_notch_width_percent": "0",
          "rc_smoothing_active_cutoffs": "60, 80",
          "acc_hardware": "0",
          "deadband": "0",
          "tpa_rate": "65",
          "rc_smoothing_cutoffs": "0, 0",
          "gyro_lowpass2_type": "0",
          "rate_limits": "1998,1998,1998",
          "dterm_notch_cutoff": "0",
          "pid_process_denom": "1",
          "gyro_cal_on_first_arm": "0",
          "yaw_lowpass_hz": "70",
          "acc_limit_yaw": "0",
          "dyn_notch_max_hz": "350",
          "yaw_deadband": "0",
          "mag_hardware": "0",
          "maxthrottle": "2000",
          "vbat_pid_gain": "0",
          "I interval": "256",
          "dterm_lowpass_dyn_hz": "84,204",
          "use_integrated_yaw": "0",
          "dshot_bidir": "1",
          "ff_boost": "0",
          "gyro_sync_denom": "1",
          "rpm_notch_lpf": "150",
          "rc_smoothing_filter_type": "1, 1",
          "serialrx_provider": "9",
          "dshot_idle_value": "550",
          "feedforward_weight": "139,147,139",
          "d_min_advance": "20",
          "rc_smoothing_debug_axis": "0",
          "gyro_rpm_notch_q": "800",
          "rc_interpolation_channels": "2",
          "dterm_filter2_type": "0",
          "ff_interpolate_sp": "3",
          "gyro_lowpass_type": "0",
          "Log start datetime": "2022-02-02T15:04:53.139+00:00",
          "dterm_filter_type": "0",
          "vbatcellvoltage": "330,350,430",
          "anti_gravity_mode": "0",
          "rc_interpolation_interval": "19",
          "rc_smoothing_type": "1",
          "iterm_relax": "1",
          "gyro_rpm_notch_harmonics": "3",
          "rc_interpolation": "2",
          "tpa_breakpoint": "1350",
          "d_min_gain": "37",
          "dterm_lowpass_hz": "150",
          "rates": "62,62,65",
          "iterm_relax_cutoff": "10",
          "d_min": "55,59,0",
          "acc_limit": "0",
          "ff_spike_limit": "55",
          "ff_max_rate_limit": "100",
          "gyro_lowpass_hz": "200",
          "gyro_notch_hz": "0,0",
          "anti_gravity_threshold": "250",
          "acc_lpf_hz": "1000",
          "baro_hardware": "0",
          "motor_pwm_protocol": "7",
          "rc_rates": "100,100,100",
          "pidAtMinThrottle": "1",
          "pitchPID": "84,126,90",
          "magPID": "40",
          "P interval": "8",
          "pidsum_limit_yaw": "1000",
          "pidsum_limit": "1000",
          "rates_type": "0",
          "Firmware date": "Jun 14 2020 03:05:04",
          "dyn_notch_min_hz": "125",
          "iterm_relax_type": "1",
          "dterm_rpm_notch_min": "100",
          "gyro_lowpass_dyn_hz": "240,600",
          "motor_pwm_rate": "480",
          "thr_mid": "50",
          "dyn_notch_q": "250",
          "dterm_notch_hz": "0",
          "feedforward_transition": "40",
          "gyro_rpm_notch_min": "100",
          "rollPID": "77,119,83",
          "thr_expo": "0",
          "looptime": "125",
          "yawPID": "83,126,0",
          "levelPID": "50,50,75",
          "rc_expo": "30,30,20",
          "dterm_rpm_notch_harmonics": "0",
          "airmode_activate_throttle": "25",
          "P ratio": "32",
          "anti_gravity_gain": "10000",
          "gyro_hardware_lpf": "0",
          "dterm_lowpass2_hz": "180",
          "use_unsynced_pwm": "0",
          "rc_smoothing_rx_average": "6722",
          "dterm_rpm_notch_q": "500",
          "iterm_windup": "70",
          "abs_control_gain": "0",
          "gyro_notch_cutoff": "0,0",
          "rc_smoothing_auto_factor": "20",
          "gyro_lowpass2_hz": "300",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 3,
          main: 16774,
          slow: 3,
          gps: 86,
          gps_home: 1,
        ),
      ),
      capped: false,
      events: [
        SyncBeep(451840837),
        Disarm(4),
        End(
          disarm_reason: None,
        ),
      ],
      main: MainSnapshot(
        count: 16774,
        last_iteration: 134149,
        time: FieldSnapshot(
          min: 0,
          max: 469230773,
          unique: 16774,
          histogram: (1070, 1082, 1113, 945, 1091, 677, 1378, 1050, 1048, 1050, 1119, 951, 1102, 684, 1362, 1052),
        ),
        fields: Fields([
          FieldSnapshot(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -49,
              max: 80,
              unique: 89,
              histogram: (3977, 3227, 2530, 1824, 1341, 960, 672, 541, 367, 292, 243, 241, 192, 184, 117, 66),
            ),
          ),
          FieldSnapshot(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -583,
              max: 284,
              unique: 198,
              histogram: (3662, 2960, 2170, 1506, 1171, 918, 826, 697, 614, 532, 421, 398, 316, 223, 181, 179),
            ),
          ),
          FieldSnapshot(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -73,
              max: 23,
              unique: 72,
              histogram: (6030, 4391, 2658, 1541, 916, 496, 243, 160, 95, 69, 49, 42, 32, 14, 22, 16),
            ),
          ),
          FieldSnapshot(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -50,
              max: 11,
              unique: 62,
              histogram: (1888, 864, 1203, 1858, 1494, 1409, 1772, 1633, 1183, 799, 506, 377, 347, 416, 483, 542),
            ),
          ),
          FieldSnapshot(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -251,
              max: 92,
              unique: 195,
              histogram: (1567, 839, 1346, 1352, 1610, 2571, 1943, 1133, 1053, 942, 765, 460, 403, 306, 258, 226),
            ),
          ),
          FieldSnapshot(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -61,
              max: 44,
              unique: 101,
              histogram: (2713, 1770, 2205, 1936, 1969, 1962, 1453, 906, 614, 476, 231, 146, 131, 104, 64, 94),
            ),
          ),
          FieldSnapshot(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -122,
              max: 192,
              unique: 106,
              histogram: (3422, 3001, 2709, 2042, 1647, 1219, 864, 605, 405, 274, 181, 124, 95, 72, 60, 54),
            ),
          ),
          FieldSnapshot(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -893,
              max: 526,
              unique: 140,
              histogram: (3571, 3196, 2698, 2045, 1537, 1019, 742, 592, 384, 265, 225, 163, 137, 85, 69, 46),
            ),
          ),
          FieldSnapshot(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1,
              max: 3,
              unique: 5,
              histogram: (16419, 312, 23, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -2,
              max: 2,
              unique: 5,
              histogram: (16269, 436, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -7,
              max: 1,
              unique: 9,
              histogram: (16539, 190, 9, 5, 4, 7, 6, 14, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -85,
              max: 98,
              unique: 184,
              histogram: (7414, 1242, 544, 800, 613, 687, 543, 740, 459, 576, 617, 375, 495, 652, 456, 561),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -76,
              max: 111,
              unique: 188,
              histogram: (985, 1079, 802, 6789, 1319, 736, 625, 703, 404, 426, 512, 361, 500, 675, 405, 453),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -96,
              max: 78,
              unique: 175,
              histogram: (1443, 4241, 1232, 1963, 1109, 883, 532, 1212, 382, 529, 531, 402, 627, 880, 407, 401),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1948,
              unique: 754,
              histogram: (950, 850, 794, 582, 880, 1324, 1449, 765, 2428, 798, 642, 1286, 1288, 986, 1004, 748),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -27,
              max: 31,
              unique: 59,
              histogram: (7432, 1193, 766, 794, 914, 1184, 696, 694, 702, 494, 374, 452, 345, 197, 292, 245),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -23,
              max: 36,
              unique: 60,
              histogram: (1359, 8165, 935, 851, 907, 857, 890, 535, 486, 322, 264, 225, 225, 218, 313, 222),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -35,
              max: 28,
              unique: 64,
              histogram: (4666, 3142, 1700, 827, 1018, 856, 562, 972, 673, 351, 248, 497, 393, 253, 283, 333),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 965,
              unique: 729,
              histogram: (2417, 759, 774, 1271, 1280, 956, 958, 686, 926, 823, 796, 690, 911, 1293, 1345, 889),
            ),
          ),
          FieldSnapshot(
            name: "vbatLatest",
            unit: Voltage,
            history: FieldSnapshot(
              min: 0,
              max: 22810,
              unique: 140,
              histogram: (1776, 0, 2268, 0, 2050, 0, 2151, 0, 2643, 0, 1976, 0, 1776, 0, 2134, 0),
            ),
          ),
          FieldSnapshot(
            name: "amperageLatest",
            unit: Amperage,
            history: FieldSnapshot(
              min: 0,
              max: 119780,
              unique: 591,
              histogram: (3455, 0, 2868, 0, 3594, 0, 1300, 0, 1245, 0, 1358, 0, 1199, 0, 1755, 0),
            ),
          ),
          FieldSnapshot(
            name: "magADC[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -740,
              max: 292,
              unique: 117,
              histogram: (1282, 791, 1282, 1677, 984, 786, 1185, 493, 1084, 591, 592, 1480, 1427, 984, 852, 1284),
            ),
          ),
          FieldSnapshot(
            name: "magADC[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -66,
              max: 947,
              unique: 126,
              histogram: (1480, 395, 985, 888, 1382, 1775, 393, 982, 1679, 951, 1476, 987, 689, 1184, 639, 889),
            ),
          ),
          FieldSnapshot(
            name: "magADC[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2576,
              unique: 123,
              histogram: (1187, 1281, 1084, 788, 493, 690, 1084, 1133, 885, 1085, 1149, 1579, 1084, 1182, 788, 1282),
            ),
          ),
          FieldSnapshot(
            name: "BaroAlt",
            unit: Unitless,
            history: FieldSnapshot(
              min: -656,
              max: 289,
              unique: 255,
              histogram: (914, 419, 1376, 979, 1014, 716, 1506, 2241, 690, 1057, 689, 839, 1078, 568, 1652, 1036),
            ),
          ),
          FieldSnapshot(
            name: "rssi",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1023,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -30,
              max: 37,
              unique: 68,
              histogram: (3820, 3984, 1450, 956, 904, 827, 664, 694, 672, 574, 486, 512, 415, 386, 236, 194),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -120,
              max: 197,
              unique: 136,
              histogram: (3222, 4965, 1926, 1212, 953, 853, 674, 705, 501, 361, 324, 194, 254, 317, 134, 179),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -43,
              max: 28,
              unique: 72,
              histogram: (3436, 3829, 1941, 1067, 995, 946, 739, 901, 532, 342, 394, 391, 383, 347, 280, 251),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -716,
              max: 566,
              unique: 387,
              histogram: (1225, 1347, 1365, 1242, 1167, 1179, 1050, 1114, 911, 921, 848, 1041, 817, 874, 876, 797),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -414,
              max: 34,
              unique: 222,
              histogram: (923, 914, 881, 921, 1025, 937, 1047, 990, 1171, 1135, 1131, 1371, 1123, 1119, 1023, 1063),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: 0,
              max: 3614,
              unique: 1680,
              histogram: (1014, 1025, 993, 1109, 1029, 1032, 1136, 1042, 1267, 1083, 1062, 976, 995, 1004, 1007, 1000),
            ),
          ),
          FieldSnapshot(
            name: "debug[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -33,
              max: 45,
              unique: 75,
              histogram: (1936, 3360, 2413, 1737, 1282, 1004, 844, 693, 642, 568, 541, 477, 373, 356, 294, 254),
            ),
          ),
          FieldSnapshot(
            name: "debug[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -125,
              max: 214,
              unique: 141,
              histogram: (2061, 3251, 2580, 1998, 1631, 1159, 906, 702, 534, 392, 362, 301, 294, 224, 222, 157),
            ),
          ),
          FieldSnapshot(
            name: "debug[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -54,
              max: 33,
              unique: 83,
              histogram: (1684, 2803, 2474, 1983, 1642, 1230, 902, 707, 604, 544, 466, 435, 380, 329, 307, 284),
            ),
          ),
          FieldSnapshot(
            name: "debug[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2047,
              unique: 1172,
              histogram: (1017, 999, 1034, 981, 1050, 1026, 1033, 1022, 1030, 1069, 1075, 1066, 1133, 1045, 1112, 1082),
            ),
          ),
          FieldSnapshot(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2044,
              unique: 1193,
              histogram: (1078, 1056, 985, 1017, 983, 1077, 1030, 1067, 1009, 989, 1020, 1103, 1023, 1192, 1083, 1062),
            ),
          ),
          FieldSnapshot(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1901,
              unique: 1153,
              histogram: (1057, 1042, 1023, 1079, 1035, 1038, 979, 1029, 1048, 1058, 1051, 1037, 1096, 1126, 1069, 1007),
            ),
          ),
          FieldSnapshot(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2052,
              unique: 1190,
              histogram: (1032, 1067, 1018, 1055, 1030, 1063, 997, 1030, 1048, 1040, 1044, 1088, 1201, 1049, 1013, 999),
            ),
          ),
        ]),
      ),
      slow: SlowSnapshot(
        count: 3,
        fields: Fields([
          FieldSnapshot(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {
              "ARM": 3,
              "BLACKBOX": 3,
            },
          ),
          FieldSnapshot(
            name: "stateFlags",
            unit: State,
            history: {
              "GPS_FIX": 3,
              "GPS_FIX_HOME": 3,
            },
          ),
          FieldSnapshot(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 3,
            },
          ),
          FieldSnapshot(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 3,
              no: 0,
            ),
          ),
          FieldSnapshot(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 3,
              no: 0,
            ),
          ),
        ]),
      ),
      gps: GpsSnapshot(
        count: 86,
        time: (),
        fields: Fields([
          FieldSnapshot(
            name: "GPS_numSat",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 8,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "GPS_coord[0]",
            unit: GpsCoordinate,
            history: FieldSnapshot(
              min: 0,
              max: 503976202,
              unique: 84,
              histogram: (9, 4, 7, 5, 2, 5, 6, 3, 5, 3, 7, 6, 8, 6, 5, 5),
            ),
          ),
          FieldSnapshot(
            name: "GPS_coord[1]",
            unit: GpsCoordinate,
            history: FieldSnapshot(
              min: 0,
              max: 74973158,
              unique: 82,
              histogram: (4, 9, 4, 13, 7, 3, 4, 6, 4, 5, 7, 1, 4, 7, 5, 3),
            ),
          ),
          FieldSnapshot(
            name: "GPS_altitude",
            unit: Altitude,
            history: FieldSnapshot(
              min: 0,
              max: 665,
              unique: 42,
              histogram: (7, 5, 4, 6, 2, 4, 8, 11, 8, 6, 3, 3, 5, 4, 5, 5),
            ),
          ),
          FieldSnapshot(
            name: "GPS_speed",
            unit: Velocity,
            history: FieldSnapshot(
              min: 0,
              max: 266,
              unique: 60,
              histogram: (2, 5, 2, 7, 8, 4, 7, 9, 4, 7, 4, 4, 8, 3, 7, 5),
            ),
          ),
          FieldSnapshot(
            name: "GPS_ground_course",
            unit: GpsHeading,
            history: FieldSnapshot(
              min: 0,
              max: 879,
              unique: 58,
              histogram: (5, 8, 4, 3, 4, 4, 3, 5, 5, 3, 3, 6, 6, 10, 1, 16),
            ),
          ),
        ]),
      ),
    )),
  ],
)
